### PR TITLE
RUMM-1814 Provide functional tests for Datadog Gradle plugin

### PR DESCRIPTION
**Pull request by mariusc83**
### What does this PR do?

In this PR we are adding functional tests for the Datadog Gradle Plugin:
- testing `assemble` task with and without the `--build-cache` option
- testing `assemble` task with and without the `--configuration-cache` option
- testing `upload` task for all the flavours
- testing the plugin configuration options

### Motivation

What inspired you to submit this pull request?

### Additional Notes

Anything else we should know when reviewing?

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] Make sure you discussed the feature or bugfix with the maintaining team in an Issue
- [ ] Make sure each commit and the PR mention the Issue number (cf the [CONTRIBUTING](CONTRIBUTING.md) doc)

### CODE REVIEW

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -99,6 +99,7 @@ test:plugin:
     - git fetch --depth=1 origin main
     - rm -rf ~/.gradle/daemon/
     - CODECOV_TOKEN=$(aws ssm get-parameter --region us-east-1 --name ci.dd-sdk-android-gradle-plugin.codecov-token --with-decryption --query "Parameter.Value" --out text)
+    - GRADLE_OPTS="-XX:MaxPermSize=512m -Xmx2560m" ./gradlew :dd-sdk-android-gradle-plugin:publishToMavenLocal -Pdd-skip-signing
```
Comment (0xnm) [inline .gitlab-ci.yml:102]: Is it necessary?

File: buildSrc/src/main/kotlin/com/datadog/gradle/config/MavenConfig.kt [hunk 2]
```diff
@@ -114,7 +114,7 @@ fun Project.publishingConfig(projectDescription: String) {
         signingExtension.apply {
             val privateKey = System.getenv("GPG_PRIVATE_KEY")
             val password = System.getenv("GPG_PASSWORD")
-            isRequired = true
+            isRequired = !hasProperty("dd-skip-signing")
```
Comment (0xnm) [inline buildSrc/src/main/kotlin/com/datadog/gradle/config/MavenConfig.kt:117]: should be reverted

File: gradle/libs.versions.toml [hunk 3]
```diff
@@ -42,7 +42,7 @@ kotlinGrammarParser = "c35b50fa44"
 
 # Datadog
 datadogSdk = "1.11.0"
-datadogPluginGradle = "1.2.0"
+datadogPluginGradle = "1.2.1-dev"
```
Comment (0xnm) [inline gradle/libs.versions.toml:45]: so to make things work we always need to have plugin in maven local?

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+    @Test
+    fun `M success W assemble`() {
+        // Given
+        stubGradleBuildFromResourceFile("build_with_datadog_dep.gradle.kts")
+
+        // When
+        val result = GradleRunner.create()
+            .withProjectDir(testProjectDir)
+            .withArguments("assemble")
+            .withPluginClasspath()
```
Comment (0xnm) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:48]: It seems there is no need to publish to the maven local if we use this method? If I get explanation here https://docs.gradle.org/current/javadoc/org/gradle/testkit/runner/GradleRunner.html#withPluginClasspath-- right

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+        assertThat(uploadTask).isNull()
+    }
+
+    // endregion
+
+    // region Internal
+
+    private fun stubFile(destination: File, content: String) {
+        with(destination.outputStream()) {
+            this.write(content.toByteArray())
```
Comment (0xnm) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:235]:
```suggestion
            write(content.toByteArray())
```

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+        const val SETTINGS_FILE_CONTENT = "rootProject.name = 'sample'"
+        val GRADLE_PROPERTIES_FILE_CONTENT = """
+           org.gradle.jvmargs=-Xmx2560m -XX:MaxPermSize=512m
+           android.useAndroidX=true
+           DD_API_KEY=fakekey
+        """.trimIndent()
+
+        @Suppress("unused")
+        @JvmStatic
+        fun uploadTasks(): Stream<Arguments> {
```
Comment (0xnm) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:277]: in case of single method argument method source can just return a list of strings per https://junit.org/junit5/docs/5.3.0-RC1/api/org/junit/jupiter/params/provider/MethodSource.html, no need to use Stream + Argument

Comment (mariusc83) [reply dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:277]: Ah cool, didn't know that ;)

File: buildSrc/src/main/kotlin/com/datadog/gradle/config/MavenConfig.kt (hunk 2 above)
Comment (mariusc83) [reply buildSrc/src/main/kotlin/com/datadog/gradle/config/MavenConfig.kt:117]: I need this here for `publishToMavenLocal` to work. Is this what you were asking ?

File: .gitlab-ci.yml (hunk 1 above)
Comment (mariusc83) [reply .gitlab-ci.yml:102]: Yes, unfortunately was not able to make it work to get the plugin from the classpath. For some reason it doesn't work with AGP. I am going to give it one more shot today.

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+            .withArguments("--build-cache", "assemble")
+            .withPluginClasspath()
+            .build()
+
+        // Then
+        assertThat(result.task(":assemble")?.outcome).isEqualTo(TaskOutcome.SUCCESS)
+    }
+
+    @Test
+    fun `M fail W assemble { configuration cache, checkProjectDependecies enabled }`() {
```
Comment (xgouchet) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:72]: Why should the plugin fail in this case? Also can we be a bit more precise in what kind of exception we're expecting ?

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+            GradleRunner.create()
+                .withProjectDir(testProjectDir)
+                .withArguments("--configuration-cache", "assemble")
+                .withPluginClasspath()
+                .build()
+        }.isInstanceOf(Exception::class.java)
+    }
+
+    // The assembleRelease fails for configuration cache and not because of our code but because of
+    // these tasks: collect[Flavour1][Flavour2]ReleaseDependencies
```
Comment (xgouchet) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:87]: Is this comment related to the test method above? If so can you move it to the test method body instead?

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+        // Given
+        stubGradleBuildFromResourceFile("build_without_datadog_dep.gradle.kts")
+        // When
+        assertThatThrownBy {
+            GradleRunner.create()
+                .withProjectDir(testProjectDir)
+                .withArguments("assemble")
+                .withPluginClasspath()
+                .build()
+        }.isInstanceOf(Exception::class.java)
```
Comment (xgouchet) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:131]: Again can we make a more precise check on which exception we're expecting here?

**Review by xgouchet — CHANGES_REQUESTED** [review]

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt (hunk 9 above)
Comment (mariusc83) [reply dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:131]: Not so sure as it is a generic exception but will try to have a look

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt (hunk 7 above)
Comment (mariusc83) [reply dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:72]: Good catch, this one should not fail actually is just that I need to test the `assembleDebug` instead as `assemble` contains `assembleRelease` also and that ones fails for other reasons than our code (see the comment below).

Comment (mariusc83) [reply dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:72]: The reason why this test should fail is that our `checkDepsTask` is not aligned with the `--configuration-cache` requirements because we are using the `project` inside. @0xnm tried to fix this a while ago but it seems too tricky for now. We have an entry in our `Readme` file that states this but I will also add an explanation into this test.

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+        // could not easily solve that.
+
+        // Given
+        stubGradleBuildFromResourceFile("build_with_datadog_dep.gradle.kts")
+
+        // When
+        assertThatThrownBy {
+            GradleRunner.create()
+                .withProjectDir(testProjectDir)
+                .withArguments("--configuration-cache", "assemble")
```
Comment (0xnm) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:87]: shouldn't we test `assembleDebug` here per test name?

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+        // Given
+        stubGradleBuildFromResourceFile("build_without_datadog_dep.gradle.kts")
+        // When
+        assertThatThrownBy {
+            GradleRunner.create()
+                .withProjectDir(testProjectDir)
+                .withArguments("assemble")
+                .withPluginClasspath(getTestConfigurationClasspath())
+                .build()
+        }.isInstanceOf(UnexpectedBuildFailure::class.java)
```
Comment (0xnm) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:137]: should we check that cause is `MissingSdkException` or it is not present in the exception thrown?

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+            .build()
+
+        // Then
+        assertThat(result.task(":$taskName")?.outcome)
+            .isEqualTo(TaskOutcome.SUCCESS)
+    }
+
+    @ParameterizedTest
+    @MethodSource("uploadTasks")
+    fun `M success W upload { checkProjectDependecies disabled }`(taskName: String) {
```
Comment (0xnm) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:182]: I think this `checkProjectDependencies` has no impact on the upload task, so we can skip tests with that maybe?

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+            .withPluginClasspath(getTestConfigurationClasspath())
+            .build()
+
+        // Then
+        assertThat(result.task(":$taskName")?.outcome)
+            .isEqualTo(TaskOutcome.SUCCESS)
+    }
+
+    @Test
+    fun `M not contain any uploadTasks W minifyNotEnabeld`() {
```
Comment (0xnm) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:217]:
```suggestion
    fun `M not contain any uploadTasks W minifyNotEnabled`() {
```

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+    @ParameterizedTest
+    @MethodSource("uploadTasks")
+    fun `M success W upload`(taskName: String) {
+        // Given
+        stubGradleBuildFromResourceFile("build_with_datadog_dep.gradle.kts")
+
+        // When
+        val result = GradleRunner.create()
+            .withProjectDir(testProjectDir)
+            .withArguments(taskName)
```
Comment (0xnm) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:153]: without calling `assembleXXX` there won't be mapping file and task will be effectively no-op according to this line https://github.com/DataDog/dd-sdk-android-gradle-plugin/blob/ba2d61297317a004efab56665257dcdbaff9e0be/dd-sdk-android-gradle-plugin/src/main/kotlin/com/datadog/gradle/plugin/DdMappingFileUploadTask.kt#L106.

Mapping Task is independent and doesn't call `assemble` by itself.

Also maybe we need to add real files/modules to the `src/main`? Not sure, but it it might be that without our files there won't be any mapping file.

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+        manifestFile = File(mainSrcDir, "AndroidManifest.xml")
+        stubFile(settingsFile, SETTINGS_FILE_CONTENT)
+        stubFile(manifestFile, ANDROID_MANIFEST_FILE_CONTENT)
+        stubFile(gradlePropertiesFile, GRADLE_PROPERTIES_FILE_CONTENT)
+    }
+
+    // region Assemble
+
+    @Test
+    fun `M success W assemble`() {
```
Comment (0xnm) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:41]: can we try to add the test (well, project structure with app module with flavors + lib module without flavors) which will check that regression of this bug https://github.com/DataDog/dd-sdk-android-gradle-plugin/issues/62 doesn't happen?

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+
+    companion object {
+        val ANDROID_MANIFEST_FILE_CONTENT = """
+            <?xml version="1.0" encoding="utf-8"?>
+            <manifest xmlns:android="http://schemas.android.com/apk/res/android"
+                      package="com.example.variants">
+
+                <application
+                    android:allowBackup="true"
+                    android:supportsRtl="true">
```
Comment (0xnm) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:272]: let's add also `exported=true` here, so that compilation succeeds for API 31 (I will handle the rest in the https://github.com/DataDog/dd-sdk-android-gradle-plugin/pull/53)

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 17]
```diff
@@ -0,0 +1,323 @@
...
+
+            </manifest>
+        """.trimIndent()
+
+        const val SETTINGS_FILE_CONTENT = """
+            pluginManagement {
+                resolutionStrategy {
+                    eachPlugin {
+                        if (requested.id.id == "com.android.application") {
+                            useModule("com.android.tools.build:gradle:7.0.3")
```
Comment (0xnm) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:290]: ideally versions here should be taken from versions catalog and be the same as we use in the plugin dependencies, but tbh I don't know how to reference them.

Maybe some property which we can inject from `buildSrc` into test environment? Same for the DD SDK version in the test build script files.

**Review by 0xnm — CHANGES_REQUESTED** [review]
Awesome work! I left few comments about how can we cover more and be more flexible in the test setup.

**Review by xgouchet — CHANGES_REQUESTED** [review]

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt (hunk 8 above)
Comment (mariusc83) [reply dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt]: Actually I just wanted to explain why we cannot provide relevant tests for `assembleRelease` . I should probably state this outside of the function when the region starts.

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt (hunk 10 above)
Comment (mariusc83) [reply dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:87]: Good catch.

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt (hunk 11 above)
Comment (mariusc83) [reply dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:137]: Could not see it in the thrown exception but will have another look.

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt (hunk 14 above)
Comment (mariusc83) [reply dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:153]: There is more than that actually, it needs to be a git versioned directory otherwise it will not work (I think you already did something like that for `GitRepositoryDetectorTest`). I will have a look into this and try to provide a dummy Application class in the `src/main`

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt (hunk 15 above)
Comment (mariusc83) [reply dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:41]: Ok, I will try to see how we can achieve this.

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt (hunk 17 above)
Comment (mariusc83) [reply dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:290]: Yeah, this is going to be tricky I guess. I am thinking that maybe we can do this in another task as there are already several other things to tackle into this one.

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt (hunk 9 above)
Comment (mariusc83) [reply dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt]: Unfortuantely I cannot get the root cause of this exception as `GradleRunner` is intercepting that but it is not passing it along :(

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt (hunk 11 above)
Comment (mariusc83) [reply dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:137]: As I explained also for Xavier's comment we are not able to get the root exception as GradleRunner is actually intercepting it under the hood but not passing it along with the result or in the thrown UnexpectedBuildFailure.

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 18]
```diff
@@ -0,0 +1,511 @@
...
+            .withPluginClasspath(getTestConfigurationClasspath())
+            .build()
+
+        // Then
+        assertThat(result.task(":samples:app:assembleDebug")?.outcome)
+            .isEqualTo(TaskOutcome.SUCCESS)
+    }
+
+    // The assembleRelease fails for configuration cache and not because of our
+    // code but because of these tasks: collect[Flavour1][Flavour2]ReleaseDependencies. For this
```
Comment (xgouchet) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:207]: What is making those tasks fail, then?

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 19]
```diff
@@ -0,0 +1,511 @@
...
+    // reason the below 2 scenarios are only tested for the `debug` build type.
+
+    @Test
+    fun `M fail W assembleDebug { configuration cache, checkProjectDependecies enabled }`() {
+
+        // This test should fail because our `DdCheckSdkDepsTask` is not following the
+        // --configuration-cache best practices (we are using the project inside) and so far we
+        // could not easily solve that.
```
Comment (xgouchet) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:215]: Wait, you mean that for now the `assembleDebug` fails but because there's a bug in our project, and not because it's expected to fail? Then we shouldn't have this test at all. We need to write prescriptive tests, not descriptive ones.
Put in other word: 
- if we want the `assembleDebug` task to fail in this specific scenario (using the `build_with_datadog_dep.gradle.kts` file and running the task with `--configuration-cache`), then yes we can keep this test;
- if we want the task to succeed, but it fails for technical reasons (untestable, non compliance on our part with the configuration cache or anything else), then either we make it succeed in the test or we do write the test for success, mark it as `@Ignored` and create a task in JIRA to make it pass.

**Review by xgouchet — CHANGES_REQUESTED** [review]

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 20]
```diff
@@ -0,0 +1,504 @@
...
+        )
+        // When
+        val result = GradleRunner.create()
+            .withProjectDir(testProjectDir)
+            .withArguments(":samples:app:assembleRelease")
+            .withPluginClasspath(getTestConfigurationClasspath())
+            .build()
+
+        // Then
+        print(result.output)
```
Comment (0xnm) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:102]: should be removed

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 21]
```diff
@@ -0,0 +1,504 @@
...
+
+        // When
+        val result = GradleRunner.create()
+            .withProjectDir(testProjectDir)
+            .withArguments(":samples:app:assembleRelease")
+            .withPluginClasspath(getTestConfigurationClasspath())
+            .build()
+
+        // Then
+        assertThat(result.task(":samples:app:assembleRelease")?.outcome)
```
Comment (0xnm) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:266]: should we try also to assert that output contains a warning?

File: dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt [hunk 22]
```diff
@@ -0,0 +1,504 @@
...
+            "build_with_datadog_dep.gradle.kts",
+            appBuildGradleFile
+        )
+        val taskName = forge.anElementFrom(uploadTasks)
+        val variantName = taskName.extractVariantName()
+
+        // When
+        val result = GradleRunner.create()
+            .withProjectDir(testProjectDir)
+            .withArguments(":samples:app:assembleRelease", taskName, "--info")
```
Comment (0xnm) [inline dd-sdk-android-gradle-plugin/src/test/kotlin/com/datadog/gradle/plugin/DdAndroidGradlePluginFunctionalTest.kt:338]: nice workaround!

**Review by 0xnm — APPROVED** [review]
I left few minor comments, but overall lgtm!

**Review by xgouchet — APPROVED** [review]
